core/jobs/batch_jobs/audit_rule_validation_jobs.py

```diff
@@ -0,0 +1,1004 @@
+            model: tuple|Any. The pair of exp and
+                opportunity models or just exp model.
+
+        Returns:
+            exp_models.ExplorationModel. Returns the exp domain object.
+        """
+        if isinstance(model, tuple):
+            try:
+                return exp_fetchers.get_exploration_from_model(model[0])
+            except Exception:
+                return None
+        else:
+            try:
+                return exp_fetchers.get_exploration_from_model(model)
+            except Exception:
+                return None
+
+    @staticmethod
+    def convert_into_model_pair(
+        models_list_pair: Tuple[
+          List[exp_models.ExplorationModel],
+          List[opportunity_models.ExplorationOpportunitySummaryModel]
+    ]) -> Tuple[
+        Optional[exp_models.ExplorationModel],
+        Optional[opportunity_models.ExplorationOpportunitySummaryModel]
+    ]:
+        """Returns the pair of exp and opportunity models.
+
+        Args:
+            models_list_pair: tuple. The pair of models list.
+
+        Returns:
+            tuple. The pair of exp and opportunity models.
+        """
+        exp_model = None
+        opportunity_model = None
+        if len(models_list_pair[0]) == 1:
+            exp_model = models_list_pair[0][0]
+        if len(models_list_pair[1]) == 1:
+            opportunity_model = models_list_pair[1][0]
+        model_pair = (exp_model, opportunity_model)
+        return model_pair
+
+    @staticmethod
+    def filter_curated_explorations(model_pair: Tuple[
+        Optional[exp_models.ExplorationModel],
+        Optional[opportunity_models.ExplorationOpportunitySummaryModel]
+    ]) -> bool:
+        """Returns whether the exp model is curated or not.
+
+        Args:
+            model_pair: tuple. The pair of exp and opportunity models.
+
+        Returns:
+            bool. Returns whether the exp model is curated or not.
+        """
+        return (model_pair[0] is not None) and (model_pair[1] is not None)
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        all_explorations = (
+            self.pipeline
+            | 'Get all ExplorationModels' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all(include_deleted=False))
+            | 'Get exploration from model' >> beam.Map(
+                self.get_exploration_from_models)
+            | 'Filter valid explorations' >> beam.Filter(
+                lambda exp: exp is not None)
+        )
+
+        combine_exp_ids_and_states = (
+            all_explorations
+            | 'Combine exp id and states' >> beam.Map(
+                lambda exp: (exp.id, exp.states, exp.created_on)
+            )
+        )
+
+        exps_with_id_and_models = (
+            self.pipeline
+            | 'Get all ExplorationModels again' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all(include_deleted=False))
+            | 'Map id and exp model' >> beam.Map(
```

Comment:
Why do you need to repeat this? You can just do `x = self.pipeline | ndb_io.GetModels(...)` and then use `x` here and above in `all_explorations`, right?

Reply:
Done